### ISSUE TITLE
ICLR-2019-Ordered Neurons: Integrating Tree Structures into Recurrent Neural Networks

### ISSUE THREAD

**Issue by BrambleXu** [BrambleXu/knowledge-graph-learning#159]
**一句话总结：**

通过对neuron排序来引入归纳偏见，使用RNN能对树结构的自然语言进行更好的建模。

> 自然语言是层级结构，由不同的成分块组成。一个大的成分结束后，里面包含有很多小成分的部分。而LSTM结构允许不同的neurons跟踪不同time scale上的信息，而且对一个有层级表示的成分进行建模的时候，并没有explicit bias. 
> 这篇文章提出，通过对neurons排序，主动添加这种inductive bias. 
> 用一个输入和forget gate来保证，如果一个neuron更新的话，这个neuron下面的neuron也会被更新。

**资源：**

- [pdf](https://arxiv.org/abs/1810.09536)
- [code](https://github.com/yikangshen/Ordered-Neurons)
- [paper-with-code](

**论文信息：**

- Author: Mila/Universite de Montr ´ eal and Microsoft Research
- Dataset: language modeling, unsupervised constituency parsing, targeted syntactic evaluation (Marvin & Linzen, 2018), and logical inference (Bowman et al.,
2015).
- 

**笔记：**

自然语言通常是树结构。但是人类在学习语言的时候，不需要这样的树结构。这就有意思了，如何把树结构整合到NN 语言模型里，是很有用的。

通过使用树结构来学习更好的representations，在DL领域受到了瞩目。

给定一个句子，有很多方法可以预测潜在的树结构。
- a supervised syntactic parser。
  - 缺点： i) few languages have comprehensive annotated data for supervised parser training; 
  - ii) in some domains, syntax rules tend to be broken (e.g. in tweets); 
  - and iii) languages change over time with use, so syntax rules may evolve

另一种方法：
- grammar induction, learning the syntactic structure from raw corpora without access to expert-labeled data, remains an open problem
  - 缺点：1 suffer from inducing a trivial structure
  - 2 encounter difficulties in training caused by learning branching policies with Reinforcement Learning (RL)
  - 3 some methods are relatively complex to implement and train

而RNN在语言模型上有很好的效果。虽然输入时序列，但是RNN结构能很好预测树结构。

在这篇文章，我们提出了ordered neurons，给RNN引入inductive bias. This inductive bias promotes differentiation of the life cycle of information stored inside each neuron:
-  high-ranking neurons will store long-term information which is kept for a large number of steps
- while low-ranking neurons will store short-term information that can be rapidly forgotten

为了防止a strict division between high-ranking and low-ranking neurons，提出了一个新的激活函数，cumulative softmax, or cumax()，这个激活函数可以allocate neurons to store long/short-term information. 

We use the cumax() function to produce a vector of master input and forget gates ensuring that when a given neuron is updated (erased), all of the neurons that follow it in the ordering are also updated (erased).

Based on the cumax() and the LSTM architecture, we have designed a new model,
ON-LSTM, that is biased towards performing tree-like composition operations。

**模型图：**

![image](https://user-images.githubusercontent.com/10768193/57591619-4c176f00-756d-11e9-8a06-0cd1503f83cd.png)

**结果**：

使用的数据和任务
Code for replicating the work in "Targeted Syntactic Evaluation of Language Models." [EMNLP 2018](https://github.com/BeckyMarvin/LM_syneval). 

**接下来要看的论文**：
